### PR TITLE
p2p: break out tests into separate crate

### PR DESCRIPTION
**Pull request by xla**
In order to explore a path forward where testing utilities as well as
the assertions for a single and across crates, this change moves the
secret connection tests into the newly introduced test crate.

### CODE REVIEW

**Review by thanethomson — COMMENTED** [review]
So I understand that having a separate test crate for cross-crate integration testing makes sense, but I'm not totally clear on the reasoning behind moving (what appear to be, at least) unit tests into a separate crate.

**Comment by xla** [comment]
@thanethomson A main source of reconsidering the testing story in Rust comes from this article: https://matklad.github.io/2021/02/27/delete-cargo-integration-tests.html

Esp. the later parts where where it goes into detail how code including unit test code is treated during compilation.

**Comment by thanethomson** [comment]
> @thanethomson A main source of reconsidering the testing story in Rust comes from this article: https://matklad.github.io/2021/02/27/delete-cargo-integration-tests.html
> 
> Esp. the later parts where where it goes into detail how code including unit test code is treated during compilation.

Interesting! So this is effectively an optimization?

It does seem like a good idea to test all publicly exposed APIs from within a single crate. I suppose then that the only tests that would be located outside of the `tests` crate would be those that test private functionality?

**Comment by xla** [comment]
> Interesting! So this is effectively an optimization?

You can look at it that way for sure, for me that is a means to an end, in this case developer experience. Where we optimise for tight feedback loops by being adamant about how quickly things fail/pass.

> It does seem like a good idea to test all publicly exposed APIs from within a single crate. I suppose then that the only tests that would be located outside of the `tests` crate would be those that test private functionality?

Exactly, I'd be surprised that there are more of a handful of cases for private functionality to be tested.

File: p2p/src/error.rs [hunk 1]
```diff
@@ -7,13 +7,13 @@ use thiserror::Error;
 pub enum Error {
     /// Cryptographic operation failed
     #[error("cryptographic error")]
-    CryptoError,
+    Crypto,
```
[the remark this answers is not in the document]
Comment (xla) [reply p2p/src/error.rs:10]: Add an entry, could you double check that this correctly formatted?

**Review by thanethomson — APPROVED** [review]

File: p2p/src/error.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (thanethomson) [reply p2p/src/error.rs:10]: Seems good - we can always tweak formatting during the next release.

**Pull request merged**